common/src/main/java/dev/latvian/mods/kubejs/core/mixin/common/ParticleEngineMixin.java

```diff
@@ -0,0 +1,18 @@
+package dev.latvian.mods.kubejs.core.mixin.common;
+
+import dev.latvian.mods.kubejs.bindings.event.ClientEvents;
+import dev.latvian.mods.kubejs.client.ParticleProviderRegistryEventJS;
+import net.minecraft.client.particle.ParticleEngine;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(ParticleEngine.class)
+public abstract class ParticleEngineMixin {
+
+	@Inject(method = "registerProviders", at = @At("RETURN"))
+	private void registerParticleProvidersKJS(CallbackInfo ci) {
+		ClientEvents.PARTICLE_PROVIDER_REGISTRY.post(new ParticleProviderRegistryEventJS());
+	}
+}
```

Comment:
Is a mixin really the best way to do this? I would think that platforms provide a way to do this.

Reply:
Huh, just tested and `ParticleType`s are registered by the time `ClientLifecycleEvents#CLIENT_STARTED` posts, which I had dismissed earlier because `ClientLifecycleEvents#CLIENT_SETUP`, which states registries are initialized by then, didn't